src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        private static bool IsLegalUsingStatementType(Compilation compilation, ITypeSymbol type)
+        {
+            var disposableType = compilation.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null)
+            {
+                return default;
+            }
+
+            // CS1674: type used in a using statement must be implicitly convertible to 'System.IDisposable'
+            return compilation.ClassifyCommonConversion(type, disposableType).IsImplicit;
+        }
+
+        private async Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var (declarationStatement, localVariable) = await FindDisposableLocalDeclaration(document, span, cancellationToken).ConfigureAwait(false);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!(declarationStatement.Parent is TBlockSyntax parent))
+            {
+                throw new NotImplementedException();
+            }
+
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+
+            var lastUsageStatement = FindSiblingStatementContainingLastUsage(
+                declarationStatement,
+                localVariable,
+                semanticModel,
+                document.GetLanguageService<ISyntaxFactsService>(),
+                cancellationToken);
+
+            var parentStatementsList = GetStatements(parent);
+
+            var declarationStatementIndex = parentStatementsList.IndexOf(declarationStatement);
+
+            var lastUsageStatementIndex = lastUsageStatement is null
+                ? -1
+                : parentStatementsList.IndexOf(lastUsageStatement, declarationStatementIndex + 1);
+
+            var statementsToSurround = GetStatementsToSurround(parentStatementsList, declarationStatementIndex, lastUsageStatementIndex);
+
+
+            var trailingTrivia = SplitTrailingTrivia(declarationStatement, syntaxFactsService);
+
+            var usingStatement =
+                CreateUsingStatement(
+                    declarationStatement.WithoutTrivia(),
+                    trailingTrivia.sameLine,
+                    statementsToSurround)
+                    .WithLeadingTrivia(declarationStatement.GetLeadingTrivia())
+                    .WithTrailingTrivia(trailingTrivia.followingLines);
+
+            var newParent = WithStatements(
+                parent,
+                ReplaceRange(
+                    parentStatementsList,
+                    index: declarationStatementIndex,
+                    count: 1 + statementsToSurround.Count,
+                    newNodes: new[] { usingStatement }));
```

Comment:
i feel like thsi is a lot more confusing than just getting an editor, and removing the sibling statements.  ymmv

Reply:
I tried SyntaxEditor and I'm not happy with the result. It feels like it ends up all implementation details, whereas this specifies more declaratively what underlying operation is happening. It lines up perfectly with `statementsToSurround`.

Can you give an example of how to make this end up simpler using SyntaxEditor?